[#69325] Reimplement "Check all"/"Uncheck all", fixing its use with Collapsible Sections

myabc opened the pull request:
# Ticket

https://community.openproject.org/wp/69325

# What are you trying to accomplish?

Fixes an issue where clicking "Check all"/"Uncheck all" opens and closes collapsible sections.

Accomplishes this by refactoring the "check all/uncheck all" functionality in OpenProject from legacy global JavaScript helpers to a modern Stimulus-based architecture. The change affects workflows and roles administration pages where users can bulk select/deselect checkboxes within collapsible sections.

## Key Changes

- Introduces two new Stimulus controllers (`CheckAllController` and `CheckableController`) that work together via outlets.
- Creates a new` CheckAllComponent` ViewComponent to encapsulate the "Check all/Uncheck all" UI
- Removes deprecated global JavaScript helpers (G`lobalHelpers.checkAll` and `GlobalHelpers.toggleCheckboxesBySelector`)
- Updates multiple view templates to use the new component and Stimulus controllers
- Adds unit tests for both the component and controllers.

## Other small improvements

- **Accessibility**: 
  - better tooltip text on Permissions Report page. "Check all/Uncheck all in row", "Check all/Uncheck all in column".
  - `aria-controls` is added to "Check all"/"Uncheck all" buttons when these buttons are not descendants of a fieldset/checkbox group container.
- **Styling**: uses a label rather than italics to indicate "built in" roles.
- **Moderation**: Removes a seemingly unused ERb template.

## Screenshots

### Permissions Report

<img width="1186" height="702" alt="image" src="https://github.com/user-attachments/assets/5d67d760-4f80-4839-9445-e8b68c897533" />

<img width="425" height="46" alt="image" src="https://github.com/user-attachments/assets/38160c13-40cb-4641-8b8e-5186ae25073c" />

### Workflow

<img width="1204" height="725" alt="Screenshot 2025-11-26 at 16 45 16" src="https://github.com/user-attachments/assets/94e1fa44-4278-4395-9d47-a957abd0e99a" />

# What approach did you choose and why?

The `CheckAllController` and `CheckableController` are designed to be robust and flexible. As such, extra care has been taken to add basic documentation and spec coverage.

## What this PR doesn't do

- It doesn't Primerize the checkbox controls on any of these pages. This would be fairly trivial to accomplish however.
- It doesn't fix truncated button text on the Workflows page: [OP #69565](https://community.openproject.org/work_packages/69565)
- It doesn't fix existing accessibility violations (missing checkbox labels): ##21288 / [OP #69564](https://community.openproject.org/work_packages/69564)

# Merge checklist

- [X] Added/updated tests
- [X] Added YARD documentation.
- [X] Added JSDOc documentation.
- [ ] Added/updated documentation in Lookbook (patterns, previews, etc)
- [X] Tested major browsers (Chrome, Firefox, Edge, ...)

myabc committed:
Remove unused member_form partial

myabc committed:
[#69325] Rework "Check all" with Stimulus

Resolves an issue when used inside collapsible section headers: clicking
"Check all"/"Uncheck all" should not toggle the collapsible section.

**Behavior**

Introduces `CheckableController` and `CheckAllController`, replacing
`GlobalHelpers` functions and `link_to_function` calls.

**Rendering**

Provides `OpenProject::Common::CheckAllComponent` to render a standard
UI and take care of connecting appropriate Stimulus controller/outlets,
actions.

myabc commented:
@copilot Taking care to keep the spec suite passing and avoid eslint errors, please add a missing example for `toggleSelection` to the `CheckableController` Jasmine spec. Test both with missing and present params.

https://stimulus.hotwired.dev/reference/actions#action-parameters

Note that the built-in Stimulus typecasting is annoying for our case case. These params should only ever be `Number`, `String` or `Boolean` - and convertible to String.

myabc commented on app/views/roles/report.html.erb:
```diff
@@ -87,15 +89,30 @@ See COPYRIGHT and LICENSE files for more details.
+                          <%= render(Primer::BaseComponent.new(tag: :div, display: :inline_flex, align_items: :center)) do %>
+                            <%= render(Primer::Beta::Text.new(mr: 1)) do %>
+                              <%= role.name %>
+                              <%= render(Primer::Beta::Label.new(scheme: :secondary)) { t(:label_builtin) } if role.builtin? %>
```
Check this styling!

myabc committed:
Apply aria-controls on non-descendant buttons

Set the `aria-control` attribute server-side in `CheckAllComponent` and
client-side in `CheckAllController` (belt and braces approach).

Also introduces `attributeTokenList` helper function to ease working
with HTML attributes that accept a space-delimited list of tokens.

myabc commented on frontend/src/stimulus/controllers/check-all.controller.ts:
```diff
@@ -0,0 +1,102 @@
+  }
+
+  checkableOutletDisconnected(_outlet:CheckableController, element:CheckableElement) {
+    attributeTokenList(this.element, 'aria-controls').remove(element.id);
+  }
+
+  /**
+   *
+   * @param event
+   * @see {@link CheckableController#checkAll}
+   */
+  checkAll(event:Event) {
+    this.checkableOutlets.forEach((outlet) => { outlet.checkAll(event); });
+  }
+
+  /**
+   *
+   * @param event
+   * @see {@link CheckableController#uncheckAll}
+   */
+  uncheckAll(event:Event) {
+    this.checkableOutlets.forEach((outlet) => { outlet.uncheckAll(event); });
+  }
+
+  /**
+   *
+   * @param event
+   * @see {@link CheckableController#toggleAll}
+   */
+  toggleAll(event:Event) {
+    this.checkableOutlets.forEach((outlet) => { outlet.toggleAll(event); });
+  }
```
@copilot open a new pull request to apply changes based on [this feedback](https://github.com/opf/openproject/pull/21134#discussion_r2566298218)

myabc committed:
Fix Types workflow feature for changed markup

bsatarnejad commented on app/components/open_project/common/check_all_component.rb:
```diff
@@ -0,0 +1,115 @@
+            data: { action: }
+          }
+        )
+        system_arguments[:aria] = merge_aria(
```
aria-controls might be empty or invalid here. When use_outlet? is false, controls is nil, aria-controls="".
Empty ARIA attributes are invalid.

It can be:

```
if use_outlet?
  system_arguments[:aria] = merge_aria(system_arguments, { aria: { controls: checkable_id } })
end
```

myabc replied:
If `checkable_id` is nil then no `aria-controls` tag is rendered. I've added some specs to verify this: https://github.com/opf/openproject/pull/21134/changes#diff-3afed4ec0d371514cfd8dcf01a1c00dad9b2a41844e9f3745f5e1f53ca321e7cR106-R116

bsatarnejad commented on app/components/open_project/common/check_all_component.rb:
```diff
@@ -0,0 +1,115 @@
+          system_arguments, { aria: { controls: } }
+        )
+
+        Primer::Beta::Button.new(scheme: :link, **system_arguments).with_content(text)
```
It is better to add a contextual label here. It isn't clear Check all what?

myabc replied:
The component renders standard button text "Check all"/"Uncheck all". I think this is probably good enough? We could an additional `aria-description`, but this would be need to be provided by the caller.

myabc replied:
I would prefer to address this in a follow up PR.

bsatarnejad commented on frontend/src/app/shared/helpers/dom-helpers.ts:
```diff
@@ -143,3 +143,77 @@ export function ensureId(el:HTMLElement, prefix = 'el'):string {
+  const setTokens = (tokens:string[]) =>
+    element.setAttribute(attribute, tokens.join(' '));
+
+  const list = Object.create(DOMTokenList.prototype) as DOMTokenList;
```
I prefer `const list = {} as DOMTokenList;` because `DOMTokenList.prototype ` In many browsers it contains internal slots that only native instances have

myabc replied:
Good point. Fixed.

bsatarnejad commented on frontend/src/stimulus/controllers/check-all.controller.spec.ts:
```diff
@@ -0,0 +1,142 @@
+  const checkAllTemplate = `
+    <div data-controller="check-all" data-check-all-checkable-outlet="#checkables">
+      <button id="check-all" data-action="check-all#checkAll">Check all</button>
+      <button id="uncheck-all" data-action="check-all#uncheckAll">Check all</button>
```
button text should be `Uncheck all`

bsatarnejad commented on frontend/src/stimulus/controllers/check-all.controller.spec.ts:
```diff
@@ -0,0 +1,142 @@
+  }
+
+  describe('without checkable controller', () => {
+    it('does nothing', () => {
```
This test always passes, even if the controller throws an exception or performs logic incorrectly.
You probably intended:
- assert that no errors were thrown
- or assert that no checkboxes changed

Right now it’s a false-positive test.

myabc replied:
> assert that no errors were thrown

This is easier since there are no checkboxes on the page.

bsatarnejad commented on frontend/src/stimulus/controllers/check-all.controller.spec.ts:
```diff
@@ -0,0 +1,142 @@
+      appendTemplate(checkableTemplate);
+      appendTemplate(checkAllTemplate);
+
+      await new Promise((resolve) => setTimeout(resolve, 0));
```
Stimulus connects controllers asynchronously, but not guaranteed in the same microtask. This sometimes works, but is not reliable. If the outlet hasn't fully connected yet, the test may intermittently fail.

```
await Promise.resolve(); 
await Promise.resolve();
```

myabc replied:
I've switched to waiting for the next paint via `requestAnimationFrame` - this is more appropriate than using `setTimeout` for DOM changes and is [also the approach Stimulus itself uses](https://github.com/hotwired/stimulus/blob/main/src/tests/cases/dom_test_case.ts#L50).

A more robust solution might involve polling for DOM changes/the existence of the controllers, but let's see how we fare with this in the interim.

I'm hoping to migrate us to Vitest ([OP #66563](https://community.openproject.org/wp/66563)), which may also change how tests perform.

bsatarnejad commented on frontend/src/stimulus/controllers/check-all.controller.ts:
```diff
@@ -0,0 +1,105 @@
+   * @see {@link CheckableController#uncheckAll}
+   */
+  uncheckAll(event:Event) {
+    this.checkableOutlets.forEach((outlet) => { outlet.uncheckAll(event); });
```
If two outlets connect to the same element, you may add: `aria-controls="foo foo"`
```
const id = ensureId(element);
const tokens = attributeTokenList(this.element, "aria-controls");
if (!tokens.contains(id)) tokens.add(id);
```

myabc replied:
`attributeTokenList` uses a `Set` under-the-hood. As such it should filter out any duplicates. See `frontend/src/app/shared/helpers/dom-helpers.spec.ts`

myabc committed:
check-all controller spec fixes: nextFrame(), etc.

- Use `requestAnimationFrame` to more-reliably await next browser
  rendering frame.
- Fix "do nothing" example with missing assertions.

klaustopher reviewed (approved):
Looked through the code and tried it out. Works as expected. I think all the aria discussions you already had with behri and I think she should maybe give the final go, but from my dide this is approved

myabc commented:
@klaustopher thanks. I'll tackle the ARIA attributes in #21288.
